File: 1721-swapping-nodes-in-a-linked-list/1721-swapping-nodes-in-a-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def swapNodes(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        stack, curr, count, first = [], head, k, None

        while curr:
            stack.append(curr)
            count -= 1
            if count == 0:
                first = curr
            curr = curr.next

        while k > 1:
            stack.pop()
            k -= 1
    
        first.val, stack[-1].val = stack[-1].val, first.val

        return head

File: 1721-swapping-nodes-in-a-linked-list/1721-swapping-nodes-in-a-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode *swapNodes(ListNode *head, int k) {
    int nodeIdx = 1;
    int length = 0;

    ListNode *ptrA = head;

    while (nodeIdx < k) {
      ptrA = ptrA->next;
      ++nodeIdx;
      ++length;
    }

    ListNode *ptrB = ptrA;

    while (ptrB != nullptr) {
      ptrB = ptrB->next;
      ++length;
    }

    ptrB = head;
    nodeIdx = 1;

    while (nodeIdx <= length - k) {
      ptrB = ptrB->next;
      ++nodeIdx;
    }

    int temp = ptrA->val;
    ptrA->val = ptrB->val;
    ptrB->val = temp;

    return head;
  }
};